src/Cardano/Prelude/GHC/Heap/Tree.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+{-|
+Module      :  Cardano.Prelude.GHC.Heap.Tree
+
+This module exports functions and data types for building and printing GHC
+heap object trees. Very useful for debugging issues involving the heap
+representation of Haskell expressions.
+
+-}
+
+module Cardano.Prelude.GHC.Heap.Tree
+  ( ClosureTreeOptions(..)
+  , RecurseCyclicClosures(..)
+  , TreeDepth(..)
+  , buildClosureTree
+  , buildAndRenderClosureTree
+  , depth
+  , renderClosure
+  , renderTree
+  )
+where
+
+import Cardano.Prelude.Base
+
+import Data.Text (pack, unpack)
+import Data.Tree (Tree(..), drawTree, levels)
+import GHC.Exts.Heap
+  (Box(..), Closure, allClosures, areBoxesEqual, asBox, getClosureData)
+import System.Mem (performGC)
+
+-- | The depth of a 'Tree'.
+data TreeDepth
+  = TreeDepth {-# UNPACK #-} !Int
+  -- ^ A specific tree depth bound.
+  | AnyDepth
+  -- ^ No tree depth bound.
+  deriving (Show)
+
+-- | Whether to recurse cyclic closures in a 'Closure' 'Tree'.
+data RecurseCyclicClosures
+  = RecurseCyclicClosures
+  -- ^ Recurse cyclic closures.
+  | NoRecurseCyclicClosures
+  -- ^ Do not recurse cyclic closures.
+  deriving (Show)
+
+-- | Options which detail how a 'Closure' 'Tree' should be constructed.
+data ClosureTreeOptions = ClosureTreeOptions
+  { ctoMaxDepth       :: !TreeDepth
+  -- ^ Construct a closure tree given a maximum depth.
+  , ctoCyclicClosures :: !RecurseCyclicClosures
+  -- ^ Whether to recurse cyclic closures while constructing a closure tree.
+  } deriving (Show)
+
+depth :: Tree a -> Int
+depth = length . levels
+
+isNegativeTreeDepth :: TreeDepth -> Bool
+isNegativeTreeDepth AnyDepth = False
+isNegativeTreeDepth (TreeDepth d)
+  | d < 0     = True
+  | otherwise = False
```

Comment:
This is pretty minor, but I wonder if it's more straightforward to terminate recursion at depth 0 (thus we would use `d <= 0` on L64), rather than depth -1. As long as the caller understands the convention it doesn't matter, but the former is more common.

Reply:
I think this change would also remove the need for [this special case match](https://github.com/input-output-hk/cardano-prelude/pull/49/files#diff-3da9276c74522d701835075401bdbf38R81).

Reply:
Will take a look at this in the morning

Reply:
Thanks! Made this change.